File: src/main/kotlin/com/xs0/gqlktx/types/gql/GFieldedType.java
package com.xs0.gqlktx.types.gql;

import com.xs0.gqlktx.schema.builder.TypeKind;
import com.xs0.gqlktx.schema.intro.GqlIntroField;
import com.xs0.gqlktx.schema.intro.GqlIntroInputValue;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public abstract class GFieldedType extends GBaseType {
    protected final Map<String, GField> fields;

    protected GFieldedType(String name, Map<String, GField> fields) {
        super(name);

        this.fields = fields; // can't check isEmpty() here, because it's populated later
    }

    public Map<String, GField> getFields() {
        return fields;
    }

    @Override
    public boolean validAsQueryFieldType() {
        return getKind() == TypeKind.INPUT_OBJECT;
    }

    protected void dumpFieldsToString(StringBuilder sb) {
        for (Map.Entry<String, GField> entry : fields.entrySet()) {
            sb.append("  ");
            entry.getValue().toString(sb);
            sb.append("\n");
        }
    }

    public List<GqlIntroField> getFieldsForIntrospection(boolean includeDeprecated) {
        if (getKind() == TypeKind.INPUT_OBJECT)
            return null;

        ArrayList<GqlIntroField> res = new ArrayList<>(fields.size());
        for (Map.Entry<String, GField> entry : fields.entrySet())
            res.add(new GqlIntroField(entry.getValue()));

        return res;
    }

    public List<GqlIntroInputValue> getInputFieldsForIntrospection() {
        if (getKind() != TypeKind.INPUT_OBJECT)
            return null;

        ArrayList<GqlIntroInputValue> res = new ArrayList<>(fields.size());
        for (Map.Entry<String, GField> entry : fields.entrySet())
            res.add(new GqlIntroInputValue(entry.getValue()));

        return res;
    }
}

File: src/main/kotlin/com/xs0/gqlktx/types/gql/GFieldedType.kt
package com.xs0.gqlktx.types.gql

import com.xs0.gqlktx.schema.intro.GqlIntroField

import java.util.ArrayList
import java.util.LinkedHashSet

abstract class GFieldedType protected constructor(
        name: String,
        val fields: Map<String, GField>, // note that fields are populated later not at construction time
        description: String?)
    : GBaseType(name, description) {

    protected var _interfaces: MutableSet<GInterfaceType> = LinkedHashSet()

    internal fun addInterface(interfaceType: GInterfaceType) {
        _interfaces.add(interfaceType)
    }

    fun getInterfaces(): Set<GInterfaceType> {
        return _interfaces
    }

    protected fun dumpFieldsToString(sb: StringBuilder) {
        for ((_, value) in fields) {
            sb.append("  ")
            value.toString(sb)
            sb.append("\n")
        }
    }

    fun getFieldsForIntrospection(includeDeprecated: Boolean): List<GqlIntroField> {
        val res = ArrayList<GqlIntroField>(fields.size)

        for ((_, value) in fields)
            if (includeDeprecated || !value.deprecated)
                res.add(GqlIntroField(value))

        return res
    }
}
